base/applications/cleanmgr/util.c

```diff
@@ -0,0 +1,1120 @@
+
+    _wcsupr(DriveArgGathered);
+
+    if (wcslen(DriveArgGathered) == 1)
+    {
+        StringCbCatW(DriveArgGathered, sizeof(DriveArgGathered), L":");
+    }
+
+    WCHAR* SingleDrive = LogicalDrives;
+    WCHAR RealDrive[ARR_MAX_SIZE] = { 0 };
+    while (*SingleDrive)
+    {
+        if (GetDriveTypeW(SingleDrive) == ONLY_PHYSICAL_DRIVE)
+        {
+            StringCchCopyW(RealDrive, sizeof(RealDrive), SingleDrive);
+            RealDrive[wcslen(RealDrive) - 1] = '\0';
+            if (wcscmp(DriveArgGathered, RealDrive) == 0)
+            {
+                StringCbCopyW(wcv.DriveLetter, sizeof(wcv.DriveLetter), DriveArgGathered);
+                break;
+            }
+        }
+            SingleDrive += wcslen(SingleDrive) + 1;
+    }
+    if (wcslen(wcv.DriveLetter) == 0)
+    {
+        LoadStringW(GetModuleHandleW(NULL), IDS_ERROR_DRIVE, TempText, _countof(TempText));
+        MessageBoxW(NULL, TempText, L"Warning", MB_OK | MB_ICONERROR);
+        return FALSE;
+    }
+    
+    return TRUE;
+}
+
+void SagesetProc(int nArgs, PWCHAR ArgReal, LPWSTR* ArgList)
+{
+    WCHAR *ValStr = NULL;
+    INT_PTR DialogButtonSelect;
+        
+    ValStr = RealStageFlag(nArgs, ArgReal, ArgList);
+        
+    DialogButtonSelect = DialogBoxParamW(GetModuleHandleW(NULL), MAKEINTRESOURCEW(IDD_SAGERUN), NULL, SagesetDlgProc, 0);
+        
+    if (DialogButtonSelect == IDCANCEL)
+        return;
+
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Old ChkDsk Files", bv.ChkDskClean))
+        return;
+    
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Temporary Files", bv.TempClean))
+        return;
+    
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\RAPPS Files", bv.RappsClean))
+        return;
+    
+    if (!RegValSet(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Recycle Bin", bv.RecycleClean))
+        return;
+}
+
+void SagerunProc(int nArgs, PWCHAR ArgReal, LPWSTR* ArgList, PWCHAR LogicalDrives)
+{
+    WCHAR *ValStr = NULL;
+    WCHAR* SingleDrive = LogicalDrives;
+    INT_PTR DialogButtonSelect;
+        
+    ValStr = RealStageFlag(nArgs, ArgReal, ArgList);
+        
+    if (RegQuery(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Old ChkDsk Files") == ALLOW_FILE_REMOVAL)
+        bv.ChkDskClean = TRUE;
+        
+    if (RegQuery(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Temporary Files") == ALLOW_FILE_REMOVAL)
+        bv.TempClean = TRUE;
+
+    if (RegQuery(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\RAPPS Files") == ALLOW_FILE_REMOVAL)
+        bv.RappsClean = TRUE;
+        
+    if (RegQuery(ValStr, L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Explorer\\VolumeCaches\\Recycle Bin") == ALLOW_FILE_REMOVAL)
```

Comment:
You probably want to free `ValStr` when you are done with it, since it is also allocated by you?